node_modules/col-activities/lib/notifications/weekly.js

```diff
@@ -0,0 +1,583 @@
+  // Translate activity type to property name.
+  var property = {
+    'asset_comment': 'comments',
+    'like': 'likes',
+    'view_asset': 'views'
+  }[activityType];
+
+  // Increment the appropriate property if a translation was found.
+  if (_.has(this, property)) {
+    this[property]++;
+  }
+};
+
+/**
+ * Given an activity type, find the appropriate running-total property and increment by one.
+ *
+ * @param  {String}    activityType     The activity type to be translated
+ */
+UserTotals.prototype.incrementActivities = function(activityType) {
+  // Translate activity type to property name.
+  var property = {
+    'get_asset_comment': 'commentsReceived',
+    'get_asset_comment_reply': 'commentsReceived',    
+    'get_like': 'likesReceived'
+  }[activityType];
+
+  // Increment the appropriate property if a translation was found.  
+  if (_.has(this, property)) {
+    this[property]++;
+  }
+};
+
+/**
+ * Given an activity type and points, find the appropriate running-total property and add points.
+ * This method is used both by CourseTotals and by UserTotals.
+ *
+ * @param  {String}    activityType     The activity type to be translated
+ * @param  {Number}    points           The number of points
+ */
+CourseTotals.prototype.incrementPoints = UserTotals.prototype.incrementPoints = function(activityType, points) {
+  var property = {
+    // Properties for activity types stored in the database.
+    'add_asset': 'pointsFromAssetsUploaded',
+    'asset_comment': 'pointsFromComments',
+    'export_whiteboard': 'pointsFromAssetsUploaded',
+    'get_asset_comment': 'pointsFromCommentsReceived',
+    'get_asset_comment_reply': 'pointsFromCommentsReceived',
+    'get_like': 'pointsFromLikesReceived',
+    'like': 'pointsFromLikes',
+    // TODO: These activity types for whiteboards are not yet implemented.
+    'whiteboard_add_asset': 'pointsFromWhiteboards',
+    'whiteboard_chat': 'pointsFromWhiteboards',
+    // Generic activity types used for summary totals.
+    'collected': 'pointsCollected',
+    'generated': 'pointsGenerated',
+    'received': 'pointsReceived'
+  }[activityType];
+
+  // Increment the appropriate property by the given number of points if a translation was found.  
+  if (_.has(this, property)) {
+    this[property] += points;
+  }
+};
+
+
+/* Utilities */
+
+
+/**
+ * Iterate over an object, subjecting each element to a transformation, and return the maximum transformed 
+ * value and one associated key (choosing the key at random in the case of a tie). If no transformation 
+ * yields a positive number, return null.
+ *
+ * @param  {Object}    object           The object to be iterated
+ * @param  {Function}  transformation   The transformation to be applied
+ * @return {Object}                     The maximum transformed value and one associated key
+ * @api private
+ */
+var sampleMaximum = function(object, transformation) {
+  var ids = null;
```

Comment:
@nicolaasmatthijs `ids` is only defined as an array once we have a nonzero value to add to it. I'll add a comment.